crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -0,0 +1,282 @@
+    builder.process_stmts(stmts);
+    builder.finish()
+}
+
+/// Control flow graph
+#[derive(Debug)]
+pub struct CFG<'stmt> {
+    /// Basic blocks - the nodes of the control flow graph
+    blocks: IndexVec<BlockId, BlockData<'stmt>>,
+    /// Entry point to the control flow graph
+    initial: BlockId,
+    /// Terminal node - will always be empty
+    terminal: BlockId,
+}
+
+impl<'stmt> CFG<'stmt> {
+    /// Index of entry point to the control flow graph
+    pub fn initial(&self) -> BlockId {
+        self.initial
+    }
+
+    /// Index of terminal node
+    pub fn terminal(&self) -> BlockId {
+        self.terminal
+    }
+
+    /// Number of basic blocks, or nodes, in the graph
+    pub fn num_blocks(&self) -> usize {
+        self.blocks.len()
+    }
+
+    /// Returns the statements comprising the basic block at the given index
+    pub fn stmts(&self, block: BlockId) -> &'stmt [Stmt] {
+        self.blocks[block].stmts
+    }
+
+    /// Returns the range of the statements comprising the basic block at the given index
+    pub fn range(&self, block: BlockId) -> TextRange {
+        self.blocks[block].range()
+    }
+
+    /// Returns the [`Edges`] going out of the basic block at the given index
+    pub fn outgoing(&self, block: BlockId) -> &Edges {
+        &self.blocks[block].out
+    }
+
+    /// Returns an iterator over the indices of the direct predecessors of the block at the given index
+    pub fn predecessors(&self, block: BlockId) -> impl ExactSizeIterator<Item = BlockId> + '_ {
+        self.blocks[block].parents.iter().copied()
+    }
+
+    /// Returns the [`BlockKind`] of the block at the given index
+    pub(crate) fn kind(&self, block: BlockId) -> BlockKind {
+        self.blocks[block].kind
+    }
+}
+
+#[newtype_index]
+pub struct BlockId;
+
+/// Holds the data of a basic block. A basic block consists of a collection of
+/// [`Stmt`]s, together with outgoing edges to other basic blocks.
+#[derive(Debug, Default)]
+struct BlockData<'stmt> {
+    kind: BlockKind,
+    stmts: &'stmt [Stmt],
+    out: Edges,
+    parents: SmallVec<[BlockId; 2]>,
+}
+
+impl Ranged for BlockData<'_> {
+    fn range(&self) -> TextRange {
+        let Some(first) = self.stmts.first() else {
+            return TextRange::default();
+        };
+        let Some(last) = self.stmts.last() else {
+            return TextRange::default();
+        };
+
+        TextRange::new(first.start(), last.end())
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+pub(crate) enum BlockKind {
+    #[default]
+    Generic,
+    /// Entry point of the control flow graph
+    Start,
+    /// Terminal node for the control flow graph
+    Terminal,
+}
+
+/// Holds a collection of edges. Each edge is determined by:
+///  - a [`Condition`] for traversing the edge, and
+///  - a target block, specified by its [`BlockId`].
+#[derive(Debug, Default, Clone)]
+pub struct Edges {
+    conditions: SmallVec<[Condition; 4]>,
```

Comment:
I don't think this has to defer this PR further but I wonder if we could use a structure similar to https://docs.rs/petgraph/latest/petgraph/graph/struct.Graph.html where `Node` and `Edge`s avoid allocations by roughly using an arena for nodes and edges and links between nodes and edges (all fixed size, using a fixed array)